### ISSUE TITLE
Running Unit Test

### ISSUE THREAD

**Issue by Gopalsuthar2308** [mbits-mirafra/axi4_avip#46]
- [x] #54
- [x] #55
- [x] #77
- [x] #80

**Comment by Gopalsuthar2308** [comment]
https://docs.svunit.org/en/latest/running_unit_tests.html

SVUnit unit tests are run using runSVUnit. there are various options of runSVUnit shown in below,

**`syntax:  runSVUnit [-s|--sim <simulator> -l|--log <log> -d|--define <macro> -f|--filelist <file> -U|-uvm -m|-mixedsim <vhdlfile>  -r|--r_arg <option> -c|--c_arg <option> -o|--out <dir> -t|--test <test> --filter <filter>]`**

1.    **-h|--help                :** prints this help screen
2.    **-s|--sim <simulator>     :** simulator is either of questa, modelsim, riviera, ius, xcelium, vcs, dsim or xsim
3.    **-l|--log <log>           :** simulation log file (default: run.log)
4.    **-d|--define <macro>      :** appended to the command line as +define+<macro>
5.    **-f|--filelist <file>     :** some verilog file list
6.    **-r|--r_arg <option>      :** specify additional runtime options
7.    **-c|--c_arg <option>      :** specify additional compile options
8.    **-U|--uvm                 :** run SVUnit with UVM
9.    **-o|--out                 :** output directory for tmp and simulation files
10.  **-t|--test                :** specifies a unit test to run (multiple can be given)
11.  **-m|--mixedsim <vhdlfile> :** consolidated file list with VHDL files and command line switches
12.  **-w|--wavedrom            :** process json files as wavedrom output
13.  **--filter <filter>     :** specify which tests to run, as <test_module>.<test_name>

**Comment by Gopalsuthar2308** [comment]
**1. -h|--help :**

`runSVUnit -h`

It will print different options and uses on terminal, shown in below,

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/c7a66ebb-3e9d-4cc4-856f-d837e8889385)

**Comment by Gopalsuthar2308** [comment]
**2. -s|--sim :** 

`runSVUnit -s <simulator>`

- it will compile and simulate existed file.
- use simulator is either of questa, modelsim, riviera, ius, xcelium, vcs, dsim or xsim.

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/00b1a9ca-809f-4423-98ee-763386f7d0b3)

**Comment by Gopalsuthar2308** [comment]
**3. -l|--log :** 

`runSVUnit -l <new_log_file> -s <simulator>`

- it will override default simulation output file run.log.

- Before using -l default run.log is there shon in below,

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/f3ce4f7e-8b22-42f7-bf62-decb041950ba)

**run command:**

`runSVUnit -l new_run.log -s questa`

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/883318a6-c2b4-4ec5-a8ac-7cc8353148d2)

- After override using -l new_run.log shown in below,

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/aae949d1-f634-464f-bbe3-d5c4fb10f1a7)

**Comment by Gopalsuthar2308** [comment]
**4. -d|--define :** 

`runSVUnit -s <simulator> -d <macro>`

It will append to the command line as `-d <macro>`

**packet_unit_test template:**

In this template I defined macro **MACRO** and value I given 5, this macro value I'm printing inside test1 and test2.

```systemverilog
`define MACRO 5
```

I declared here test1 and test2.

```systemverilog
  `SVUNIT_TESTS_BEGIN

  `SVTEST(Test1)
    $display("inside Test1 code printing");
    $display("inside Test1 code printing MACRO = %0d",`MACRO);
  `SVTEST_END

  `SVTEST(Test2)
    $display("inside Test2 code printing");
    $display("inside Test2 code printing MACRO = %0d",`MACRO);
  `SVTEST_END

  `SVUNIT_TESTS_END
```

If you want to see full packet_unit_test template just click below unit_test,
<details><summary>Unit_test</summary>
<p>

```systemverilog
`include "svunit_defines.svh"
`include "packet.sv"
`define MACRO 5
module packet_unit_test;
  import svunit_pkg::svunit_testcase;

  string name = "packet_ut";
  svunit_testcase svunit_ut;


  //===================================
  // This is the UUT that we're
  // running the Unit Tests on
  //===================================
  packet my_packet;


  //===================================
  // Build
  //===================================
  function void build();
    svunit_ut = new(name);

    my_packet = new(/* New arguments if needed */);
  endfunction


  //===================================
  // Setup for running the Unit Tests
  //===================================
  task setup();
    svunit_ut.setup();
    /* Place Setup Code Here */
    $display("setup code printing");
  endtask


  //===================================
  // Here we deconstruct anything we
  // need after running the Unit Tests
  //===================================
  task teardown();
    svunit_ut.teardown();
    /* Place Teardown Code Here */
    $display("teardown code printing");
  endtask

  `SVUNIT_TESTS_BEGIN


  `SVTEST(Test1)
    $display("inside Test1 code printing");
    $display("inside Test1 code printing MACRO = %0d",`MACRO);
  `SVTEST_END

  `SVTEST(Test2)
    $display("inside Test2 code printing");
    $display("inside Test2 code printing MACRO = %0d",`MACRO);
  `SVTEST_END

  `SVUNIT_TESTS_END

endmodule
```

</p>
</details> 


**run command:**

`runSVUnit -t packet_unit_test.sv -s questa -d MACRO=10`

- Here I'm running command and new value 10 giving for MACRO,
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/20ce1369-dd60-4e7f-8892-af3824c68265)

Here I'm using **-t** because in this directory I have multiple files but here I'm running only packet_unit_test. If you want to see explanation of **-t** follow below link,
https://github.com/mbits-mirafra/axi4_avip/issues/46#issuecomment-1781767482


**Output:**
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/523dfa59-05b8-4d4e-a53b-70ac088991b9)


Follow below link fore more about -define
- https://github.com/mbits-mirafra/axi4_avip/issues/80

**Comment by Gopalsuthar2308** [comment]
**5. -f|--filelist :** 

`runSVUnit -f <filelist> -s questa`

If you declared any **.f** file then use this option.

**run command:**

`runSVUnit -f even_odd.f -s questa`

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/c277456c-eb74-4a2c-bf75-9c7d5c6faa2e)


**- Click on below example for more clarity with example.**

<details><summary>Example</summary>
<p>

<img width="773" alt="filelist" src="https://github.com/mbits-mirafra/axi4_avip/assets/110448382/d8cacf29-6d92-429a-afb8-43b566ed02e0">

Here i declared two directory one for UUT and other for unit test template, shown in below

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/1eb6e254-4188-4efe-9128-4fb749ed3d0e)
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/4025b4d5-5a61-4e78-a9c0-81df5d53f50b)
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/d57dc9cd-7b1b-4958-9a03-22cec7ce74eb)


**even_odd.sv class:**

In this class i declared variable a and wrote constraint for that. and two function defined one is_even and second is_odd.

```systemverilog
class even_odd;
  rand bit [3:0] a;

  constraint c1{a == 5;}
  function bit [3:0] is_even();
    if(a%2==0)
      return 1;
    else
      return 0;
  endfunction

  function bit [3:0] is_odd();
    if(a%2!=0)
      return 1;
    else
      return 0;
  endfunction
endclass
```

**even_odd_unit_test:**

Here i declared two testcases one is_even_true it will set status FAILED when given condition is not satisfied and second is_odd_false it will set status FAILED when given condition is not satisfied.

```systemverilog
`include "svunit_defines.svh"
`include "even_odd.sv"

module even_odd_unit_test;
  import svunit_pkg::svunit_testcase;

  string name = "even_odd_ut";
  svunit_testcase svunit_ut;


  //===================================
  // This is the UUT that we're
  // running the Unit Tests on
  //===================================
  even_odd uut;


  //===================================
  // Build
  //===================================
  function void build();
    svunit_ut = new(name);

    uut = new(/* New arguments if needed */);
  endfunction

  //===================================
  // Setup for running the Unit Tests
  //===================================
  task setup();
    svunit_ut.setup();
    /* Place Setup Code Here */

  endtask


  //===================================
  // Here we deconstruct anything we
  // need after running the Unit Tests
  //===================================
  task teardown();
    svunit_ut.teardown();
    /* Place Teardown Code Here */

  endtask
  //===================================
  // All tests are defined between the
  // SVUNIT_TESTS_BEGIN/END macros
  //
  // Each individual test must be
  // defined between `SVTEST(_NAME_)
  // `SVTEST_END
  //
  // i.e.
  //   `SVTEST(mytest)
  //     <test code>
  //   `SVTEST_END
  //===================================
  `SVUNIT_TESTS_BEGIN
    void'(my_even_odd.randomize());

  //===================================
  //  creating test case is_even_true
  //===================================
    `SVTEST(is_even_true)
       $display("a inside is_even_true test = %0d",uut.a);
       `FAIL_UNLESS(uut.is_even() === 1)
    `SVTEST_END

  //===================================
  //  creating test case is_odd_false
  //===================================
    `SVTEST(is_odd_false)
       $display("a inside is_odd_false test = %0d",uut.a);
       `FAIL_UNLESS(uut.is_odd() === 0)
    `SVTEST_END

  `SVUNIT_TESTS_END

  endmodule

```


**even_odd.f:**
Here giving path for UUT directory and files.

```systemverilog
+incdir+../UUT/
 ../UUT/*.sv
```

**run command:** 

Without even_odd.f file running inside uvm test template.

`runSVUnit -s questa`

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/1cfb5cf4-df50-48d7-a79f-7b4011f666ae)

**Output:**
Without **`.f`** file it will give error shown below,

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/9533a74d-98da-4a73-8298-7de4af2edf26)


**run command:** 

With even_odd.f file running inside uvm test template.

`runSVUnit -f even_odd.f -s questa`

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/0f557f6e-9cc6-451b-b8f1-c49120fd2689)


**Output:**

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/bd6ef838-223b-4326-8c76-549851e9157f)


</p>
</details>

**Comment by Gopalsuthar2308** [comment]
**6. -r|--r_arg :** 

- If you want to give any run time argument then you this option.

`runSVUnit -s questa -r "argument"`

> **Warning:** Use **" "** before and after argument, otherwise it will give error.



Here I am using simple model unit test template click on below unit test for full code.

<details><summary>simple_model_unit_test</summary>
<p>

```systemverilog
`include "svunit_defines.svh"

`include "simple_model.sv"
import svunit_uvm_mock_pkg::*;

class simple_model_uvm_wrapper extends simple_model;

   `uvm_component_utils(simple_model_uvm_wrapper)

   // Add some extra ports and FIFOs in order to verify the simple_model wrapper's
   // connect_phase().
   uvm_blocking_put_port #(simple_xaction) in_put_port;
   uvm_tlm_fifo #(simple_xaction)          in_put_fifo;

   uvm_tlm_fifo #(simple_xaction)          out_get_fifo;
   uvm_blocking_get_port #(simple_xaction) out_get_port;

   function new(string name = "simple_model_uvm_wrapper", uvm_component parent);
      super.new(name, parent);
      in_put_port = new("in_put_port", null);
      in_put_fifo = new("in_put_fifo", null);
      out_get_fifo = new("out_get_fifo", null);
      out_get_port = new("out_get_port", null);
   endfunction

   //===================================
   // Build
   //===================================
   function void build_phase(uvm_phase phase);
      super.build_phase(phase);
      /* Place Build Code Here */

   endfunction

   //==================================
   // Connect
   //=================================
   function void connect_phase(uvm_phase phase);
      super.connect_phase(phase);

      /* Place Connection Code Here */

      // Connect the extra ports/FIFOs to the simple_model ports.
      get_port.connect(in_put_fifo.get_export);
      in_put_port.connect(in_put_fifo.put_export);

      out_get_port.connect(out_get_fifo.get_export);
      put_port.connect(out_get_fifo.put_export);
   endfunction // connect_phase

endclass // simple_model_uvm_wrapper


module simple_model_unit_test;
   import svunit_pkg::svunit_testcase;

   string name = "simple_model_ut";
   svunit_testcase svunit_ut;


   //===================================
   // This is the UUT that we're
   // running the Unit Tests on
   //===================================
   simple_model_uvm_wrapper my_simple_model;

   //-------------------------------------------
   // for testing purposes, add fifos and ports
   // to interact with the simple_model
   //-------------------------------------------

   //===================================
   // Build
   //===================================
   function void build();
      svunit_ut = new(name);

      my_simple_model = simple_model_uvm_wrapper::type_id::create("", null);

      if (my_simple_model == null) $display("Test model null.");

      svunit_deactivate_uvm_component(my_simple_model);
   endfunction


   //===================================
   // Setup for running the Unit Tests
   //===================================
   task setup();
      svunit_ut.setup();
      /* Place Setup Code Here */

      svunit_activate_uvm_component(my_simple_model);

      //-----------------------------
      // start the testing phase
      //-----------------------------
      svunit_uvm_test_start();
   endtask


   //===================================
   // Here we deconstruct anything we
   // need after running the Unit Tests
   //===================================
   task teardown();
      svunit_ut.teardown();
      //-----------------------------
      // terminate the testing phase
      //-----------------------------
      svunit_uvm_test_finish();

      /* Place Teardown Code Here */

      // After every test, flush the fifos so leftover FIFO contents doesn't affect
      // the next test.
      my_simple_model.out_get_fifo.flush();
      my_simple_model.in_put_fifo.flush();

      svunit_deactivate_uvm_component(my_simple_model);
   endtask


   //===================================
   // All tests are defined between the
   // SVUNIT_TESTS_BEGIN/END macros
   //
   // Each individual test must be
   // defined between `SVTEST(_NAME_)
   // `SVTEST_END
   //
   // i.e.
   //   `SVTEST(mytest)
   //     <test code>
   //   `SVTEST_END
   //===================================
   `SVUNIT_TESTS_BEGIN

     //************************************************************
     // Test:
     //   get_port_not_null_test
     //
     // Desc:
     //   test for the existance of the simple_model::get_port
     //************************************************************
     `SVTEST(get_port_not_null_test)

       `FAIL_IF(my_simple_model.get_port == null);

     `SVTEST_END


      //************************************************************
      // Test:
      //   get_port_active_test
      //
      // Desc:
      //   ensure that objects put to the input do not remain in the
      //   input FIFO.
      //************************************************************
     `SVTEST(get_in_port_active_test)
       begin
         simple_xaction tr = new();

         my_simple_model.in_put_port.put(tr);
         #1;
         `FAIL_UNLESS(my_simple_model.in_put_fifo.is_empty());
       end
     `SVTEST_END

      //************************************************************
      // Test:
      //   get_port_active_test
      //
      // Desc:
      //   ensure that objects put to the input move through the input
      //   FIFO to the output FIFO.
      //************************************************************
     `SVTEST(get_out_port_active_test)
       begin
         simple_xaction tr = new();

         my_simple_model.in_put_port.put(tr);
         #1;
         `uvm_info("simple_model_unit_test", $sformatf("out_get_fifo empty : %0d", my_simple_model.out_get_fifo.is_empty()), UVM_HIGH)
         `FAIL_IF(my_simple_model.out_get_fifo.is_empty());
       end
     `SVTEST_END

     //************************************************************
     // Test:
     //   xformation_test
     //
     // Desc:
     //   ensure that objects going through the simple model have
     //   their field property updated appropriately (multiply by 2)
     //************************************************************
     `SVTEST(xformation_test)
       begin
         simple_xaction in_tr = new();
         simple_xaction out_tr;

         void'(in_tr.randomize() with { field == 2; });
         `FAIL_UNLESS(my_simple_model.out_get_fifo.is_empty());

         my_simple_model.in_put_port.put(in_tr);
         my_simple_model.out_get_port.get(out_tr);

         `FAIL_IF(in_tr.field != 2);
         `FAIL_IF(out_tr.field != 4);
       end
     `SVTEST_END

     `SVUNIT_TESTS_END

endmodule
```


</p>
</details> 


```systemverilog
     `SVTEST(get_out_port_active_test)
       begin
         simple_xaction tr = new();

         my_simple_model.in_put_port.put(tr);
         #1;
         `uvm_info("simple_model_unit_test", $sformatf("out_get_fifo empty : %0d", my_simple_model.out_get_fifo.is_empty()), UVM_HIGH)
         `FAIL_IF(my_simple_model.out_get_fifo.is_empty());
       end
     `SVTEST_END

```

Here In uvm info i'm using UVM_HIGH, By default in uvm verbosity level is UVM_MEDIUM so if I want to print this info then I need to give command line argument shown in below,

**run command:**

`runSVUnit -s questa -r "+UVM_VERBOSITY=UVM_HIGH"`

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/ea70d19a-d3c2-43b6-8b1c-bda87da67583)

**Output:**
<img width="959" alt="vervosity" src="https://github.com/mbits-mirafra/axi4_avip/assets/110448382/72b826e6-15e2-47cb-a2d7-cd830c686f19">

So in above output we can able to see uvm_info because here we set verbosity as UVM_HIGH using command line.


**- Follow the below link for uses of **-r** option:**
- https://github.com/mbits-mirafra/axi4_avip/issues/77

**Comment by Gopalsuthar2308** [comment]
**7. -c|--c_arg :** 
 
If you want to give any compile time argument then you this option.

`runSVUnit -s questa -c "argument"`

> **Warning:** Use **" "** before and after argument, otherwise it will give error.


**Click on below single argument for more clarity with example.**


<details><summary>Single argument</summary>
<p>

**Command:**

`runSVUnit -s questa -c "+define+MACRO=40"`


I declared here one packet.sv class and one unit test template.

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/0f057d8d-44a3-4a6b-a534-e67c558a4b63)

**packet_unit_test template:**

In this template I defined two macro **MACRO**  and **WIDTH** and value I given MACRO 5 and WIDTH 32, this two macros value i'm printing inside test1 and test2.

```systemverilog
`define MACRO 5
`define WIDTH 32
```

I declared here test1 and test2
```systemverilog
  `SVUNIT_TESTS_BEGIN

 //===================================
 // creating Test1
 //===================================
  `SVTEST(Test1)
    $display("inside Test1 code printing MACRO = %0d and WIDTH = %0d",`MACRO,`WIDTH);
  `SVTEST_END
 //===================================
 // creating Test2
 //===================================
  `SVTEST(Test2)
    $display("inside Test2 code printing MACRO = %0d and WIDTH = %0d",`MACRO,`WIDTH);
  `SVTEST_END

  `SVUNIT_TESTS_END
```


If you want to see full packet_unit_test template just click below unit_test,

<details><summary>packet_unit_test</summary>
<p>

```systemverilog
`include "svunit_defines.svh"
`include "packet.sv"
//declaring `define macro
`define MACRO 5
`define WIDTH 32

module packet_unit_test;
  import svunit_pkg::svunit_testcase;

  string name = "packet_ut";
  svunit_testcase svunit_ut;

  //===================================
  // This is the UUT that we're
  // running the Unit Tests on
  //===================================
  packet my_packet;

  //===================================
  // Build
  //===================================
  function void build();
    svunit_ut = new(name);
    my_packet = new(/* New arguments if needed */);
  endfunction

  //===================================
  // Setup for running the Unit Tests
  //===================================
  task setup();
    svunit_ut.setup();
    /* Place Setup Code Here */
    $display("setup code printing");
  endtask

  //===================================
  // Here we deconstruct anything we
  // need after running the Unit Tests
  //===================================
  task teardown();
    svunit_ut.teardown();
    /* Place Teardown Code Here */
    $display("teardown code printing");
  endtask

  //===================================
  // All tests are defined between the
  // SVUNIT_TESTS_BEGIN/END macros
  //
  // Each individual test must be
  // defined between `SVTEST(_NAME_)
  // `SVTEST_END
  // i.e.
  //   `SVTEST(mytest)
  //     <test code>
  //   `SVTEST_END
  //===================================
  `SVUNIT_TESTS_BEGIN

 //===================================
 // creating Test1
 //===================================
  `SVTEST(Test1)
    $display("inside Test1 code printing MACRO = %0d and WIDTH = %0d",`MACRO,`WIDTH);
  `SVTEST_END

 //===================================
 // creating Test2
 //===================================
  `SVTEST(Test2)
    $display("inside Test2 code printing MACRO = %0d and WIDTH = %0d",`MACRO,`WIDTH);
  `SVTEST_END

  `SVUNIT_TESTS_END

endmodule

```

</p>
</details> 


Now how to override MACRO using command line, so using **-c** compile time argument option we can replace macro, shown in below.

**run command:**

`runSVUnit -s questa -c "+define+MACRO=40"`

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/a9754a38-8dcc-430e-a2d8-95827e4c768f)

**Output:**
<img width="959" alt="single_arg" src="https://github.com/mbits-mirafra/axi4_avip/assets/110448382/64d02e78-1df1-473d-b333-7abef8d8317a">


</p>
</details> 

**You can give multiple argument, Click on below Multiple arguments for example,**

<details><summary>Multiple argument</summary>
<p>

**run command:**

`runSVUnit -s questa -c "+define+MACRO=40 +define+WIDTH=64"`

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/982cd1f0-b32e-43c9-a054-eee7ccbe17d8)


**Output:**

<img width="959" alt="multi_arg" src="https://github.com/mbits-mirafra/axi4_avip/assets/110448382/22ec0d6d-fce6-4b24-b40b-f60f814c727b">


</p>
</details> 



**Difference between **-c** and **-d** option**

Click on below difference,

<details><summary>Difference</summary>
<p>

**Command:**

`runSVUnit -s questa -d MACRO=30 -d WIDTH=64`

`runSVUnit -s questa -c "+define+MACRO=30 +define+MACRO=64"`

1. In `-d` option you need write only **-d** for each Macro, But in the `-c` option you need to write +define+ for each macro, so we can say `-d` option is less verbose because less typing when compare to `-c` option.

2. In `-d` option you need write **-d** for each Macro, But in the `-c` option you need to write **-c** only once for all arguments.

3. In `-d` option only macros we can pass, But in the `-c` option we can pass macros and you can pass any other compilation argument like +cover etc...

**Conclusion:**
So if you need to pass argument as a MACROS then use `-d` and if you need to pass other argument along with MACROS then use `-c`.

</p>
</details>

**Comment by Gopalsuthar2308** [comment]
**8. -U|--uvm :** 

`runSVUnit -U -s <simulator>` or 
`runSVUnit --uvm -s <simulator>`

- it will run SVUnit with UVM.
- if you have uvm file and if you run without **-uvm** then it will give error, so use **-uvm** for uvm unit test.

- follow the below link for uvm simple example,
https://github.com/mbits-mirafra/axi4_avip/issues/65

**Comment by Gopalsuthar2308** [comment]
**9. -o|--out :** 

`runSVUnit -s <simulator> -o <output_fale_name>`

It will create output directory and tmp and simulaton files will write inside that.

run command:

`runSVUnit -s questa -o output_files`

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/1a6c2a1b-2fbe-4228-9d0b-783af045780b)

output:
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/bcdbb54f-424f-443d-b439-96d79e4de6cc)

Always use -o/--out [dir] when calling runSVUnit or buildSVUnit. Both will generate a bunch of files — some hidden — in the current directory if you run them without a specific output directory.

**Comment by Gopalsuthar2308** [comment]
**10. -t|--test :** 

`runSVUnit -t <unit_test_template> -t <unit_test_template> -s <simulator>`

It will run particular unit test template or we can specifies a multiple unit test template. 

**run command:**

`runSVUnit -t packet_unit_test.sv -t even_odd_unit_test.sv -s questa`

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/92baed1b-d38d-4aa1-8b06-6c3e8cbc76d7)

It will run only packet unit test and even odd unit test shown in below,

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/4dc5a6b1-232d-47cd-b1ec-00d2cffaccdd)
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/1aa110fd-c166-4cc7-99fc-e350a7011740)

**Comment by Gopalsuthar2308** [comment]
**11. -m|--mixedsim :** 

consolidated file list with VHDL files and command line switches

**Comment by Gopalsuthar2308** [comment]
**12. -w|--wavedrom :** 

`runSVUnit -w <simulator>`

process json files as wavedrom output

- More info #64

**Comment by Gopalsuthar2308** [comment]
**13. --filter :** 

`runSVUnit --filter <unit_test_template_name>.<test_name>`

- specify which tests to run, as <test_module>.<test_name>, if you have multiple tests in single testcase and if you want to run only one test then use this option.
- in below figure i have even_odd_unit_test and even_odd class so my testcase name even_odd_ut and one test name is_even_true so i shown below, 

**run command:**

`runSVUnit --filter even_odd_ut.is_even_true`

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/9af4ed6c-5529-4d0d-a5a3-721e3092dfe7)

**follow below link fore more clarity of --filter and how to run only one test in a testcase with example,**

- https://github.com/mbits-mirafra/axi4_avip/issues/55


**follow below link fore more clarity of --filter and how to run more than one testcase (each testcase has multiple tests)**

- #87

**Comment by muneebullashariff** [comment]
Show example as how to pass more than one MACRO using the commandline

https://github.com/mbits-mirafra/axi4_avip/issues/46#issuecomment-1781761921

**Comment by muneebullashariff** [comment]
https://github.com/mbits-mirafra/axi4_avip/issues/46#issuecomment-1781762735
Create a new task and show the usage of -f in real usecase scenario

**Comment by muneebullashariff** [comment]
https://github.com/mbits-mirafra/axi4_avip/issues/46#issuecomment-1781762735
See if we can use -c instead of -d, and what is the actual difference between them

**Comment by muneebullashariff** [comment]
We also have below option. You can check that by using the help option 
  print "     --directory           : only run svunit discovery on selected     directories\n";

**Comment by Gopalsuthar2308** [comment]
**14. --directory :** 

`runSVUnit --directory <directory_name> -s <simulator>`

- if you have multiple directory and if you want to run only single directory then use this option.

Here inside structure i have four directory shown in below,
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/dc528d58-0e9b-4dc2-a055-1c078ff3dab3)

check_test have 2 files:
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/d6a5478d-292e-4586-ace4-a7edc690fcab)

multi_testrunner have two directories:
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/45125801-0e63-4b50-82bf-2d5aab24b904)

svUnit_test have 5 files:
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/e9319d24-d78c-4944-a1b0-5bbb5c951cd5)

testrunner1 have two directories:
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/3407aed7-5433-4bec-a1cd-675cbd483737)


- If i'll run `runSVUnit -s questa` then it will execute all files in this four directories, shown below,

**run command:**
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/4c52e733-9aea-4db1-b8c2-6d7574094b0c)


**Output:**
**Expand below details for output:**

<details><summary>Details</summary>
<p>

![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/836f4473-74a0-46b5-8e1f-25bc8cfda60d)
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/6725bfc5-fa53-41d5-8a1d-c51ced3488bd)
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/5515ce42-59d6-4cc4-96ed-08f06b6fc79d)
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/761e7039-d39b-4b94-a247-fddacfdbeecb)
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/1632ab37-92af-40b9-849e-c39c7971f257)
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/687f200a-6fc9-40ec-b651-12c6ffd55580)
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/74a3a826-59fd-46ff-bd5f-a15368184db7)

</p>
</details> 


- If i'll run `runSVUnit --directory  check_test -s questa` then it will execute all files in this check_test directory only, shown below,

**run command:**
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/539d6dd4-7fbc-4a12-b2d1-d1f0231d21ef)


**Output:**
![image](https://github.com/mbits-mirafra/axi4_avip/assets/110448382/7399364f-be82-485a-8953-be846df729fa)

**Comment by muneebullashariff** [comment]
**14. Waveform generation**

`runSVUnit --waves `

Used for generating the waveforms for post-processing. Currently support for questa/modelsim.

**run command:**

`runSVUnit -s questa --waves`   


Check this for more information
- #85
